var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py

```diff
@@ -67,41 +67,54 @@ class IntelOneapiMkl(IntelOneApiLibraryPackage):
     def component_dir(self):
         return 'mkl'
 
-    def xlp64_lib(self, lib):
-        return lib + ('_ilp64'
-                      if '+ilp64' in self.spec
-                      else '_lp64')
-
     @property
     def headers(self):
-        include_path = join_path(self.component_path, 'include')
-        return find_headers('*', include_path)
-
-    # provide cluster libraries if +cluster variant is used or
-    # the scalapack virtual package was requested
-    def cluster(self):
-        return '+cluster' in self.spec
+        return find_headers('*', self.component_prefix.include)
 
     @property
     def libs(self):
         shared = '+shared' in self.spec
-        mkl_libs = []
-        if self.cluster():
-            mkl_libs += [self.xlp64_lib('libmkl_scalapack'),
-                         'libmkl_cdft_core']
-        mkl_libs += [self.xlp64_lib('libmkl_intel'),
-                     'libmkl_sequential',
-                     'libmkl_core']
-        if self.cluster():
-            mkl_libs += [self.xlp64_lib('libmkl_blacs_intelmpi')]
-        libs = find_libraries(mkl_libs,
-                              join_path(self.component_path, 'lib', 'intel64'),
-                              shared=shared)
+
+        libs = self._find_mkl_libs(shared)
+
         system_libs = find_system_libraries(['libpthread', 'libm', 'libdl'])
         if shared:
             return libs + system_libs
         else:
             return IntelOneApiStaticLibraryList(libs, system_libs)
 
+    def setup_run_environment(self, env):
+        super(IntelOneapiMkl, self).setup_run_environment(env)
+
+        # Support RPATH injection to the library directories when the '-mkl' or '-qmkl'
+        # flag of the Intel compilers are used outside of the Spack build environment.
+        # We should not try to take care of other compilers because the users have to
+        # provide the linker flags anyway and are expected to take care of the RPATHs
+        # flags too. We prefer the POST flags over the PRE ones to allow for overriding
+        # of the injected RPATHs. Also, we use append_path instead of append_flags
+        # because the latter is not yet supported in the default modulefile templates.
+        for d in self._find_mkl_libs('+shared' in self.spec).directories:
+            flag = '-Wl,-rpath,{0}'.format(d)
+            env.append_path('__INTEL_POST_CFLAGS', flag, separator=' ')
+            env.append_path('__INTEL_POST_FFLAGS', flag, separator=' ')
```

Comment:
1. There are `__INTEL_PRE_CFLAGS` and `__INTEL_POST_FFLAGS` (see [here](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-dpcpp-cpp-compiler-dev-guide-and-reference/top/compilation/supported-environment-variables.html)). In the comment, I tried to explain why we should use the latter.
2. We should not `env.set` because the user might have `__INTEL_POST_FFLAGS` set to something already before loading the module and we want just to extend the value.
3. I mean both `%intel` and `%oneapi` and the modifications to `intel-oneapi-compilers` do not set these RPATHs because `intel-oneapi-compilers` is a separate package and doesn't know where `intel-oneapi-mkl` is installed (if at all).

Sorry for the confusion.

Reply:
Oh I think I see: here you mean override in the sense that if the user places other RPATHs anywhere in the options, that they will come before the ones inserted with this mechanism, and therefore take precedence. Can you clarify that in the comment?

Also, according to https://lmod.readthedocs.io/en/latest/051_tcl_modulefiles.html, TCL at least doesn't appear to have an "append-flags" function (the same appears to be true for [lmod](https://lmod.readthedocs.io/en/latest/050_lua_modulefiles.html)):

* If they in fact don't, can you rephrase the comment (if this is the case, then I think you did it the way the would have intended and there is no problem with this)?
* If they do, can you move this part of the comment to a TODO comment specifically on these`append_path` calls?

Reply:
@scheibelp I have updated the comment related to the choice of `POST` flags and removed the one about `append_path`. What I actually meant in the latter was that we could add support for `append_flags` in the module file generation via the `append-path` function with the space as delimiter.